File: R/hypotheses_joint.R
joint_test <- function(
    object,
    joint_index = NULL,
    hypothesis = 0,
    joint_test = "f",
    df = NULL,
    vcov = TRUE) {
    checkmate::assert_choice(joint_test, c("f", "chisq"))

    mfx <- components(object, "all")

    if (joint_test == "f") {
        checkmate::assert_numeric(df, len = 2, null.ok = TRUE)
    } else {
        checkmate::assert_numeric(df, len = 1, null.ok = TRUE)
    }

    # theta_hat: P x 1 vector of estimated parameters
    if (inherits(object, c("slopes", "comparisons"))) {
        nam <- object$term
        if ("contrast" %in% names(object)) {
            nam <- paste(nam, object$contrast)
        }
        theta_hat <- stats::setNames(object$estimate, nam)
    } else {
        theta_hat <- get_coef(object)
    }

    # index
    checkmate::assert(
        checkmate::check_integerish(
            joint_index,
            lower = 1,
            upper = length(theta_hat)
        ),
        checkmate::check_character(joint_index),
        checkmate::check_true(joint_index)
    )

    if (isTRUE(joint_index)) {
        joint_index <- seq_along(theta_hat)
    } else if (isTRUE(checkmate::check_string(joint_index))) {
        joint_index <- grep(joint_index, names(theta_hat), perl = TRUE)
    }

    # V_hat: estimated covariance matrix
    V_hat <- get_vcov(object, vcov = vcov)

    # R: Q x P matrix for testing Q hypotheses on P parameters
    # build R matrix based on joint_index
    R <- matrix(0, nrow = length(joint_index), ncol = length(theta_hat))
    for (i in seq_along(joint_index)) {
        if (is.numeric(joint_index)) {
            R[i, joint_index[i]] <- 1
        } else {
            R[i, which(names(theta_hat) == joint_index[i])] <- 1
        }
    }

    # null hypothesis
    checkmate::assert(
        checkmate::check_number(hypothesis),
        checkmate::check_numeric(hypothesis, len = nrow(R)),
        checkmate::check_null(hypothesis)
    )
    if (is.null(hypothesis)) {
        hypothesis <- 0
    }
    r <- matrix(hypothesis, nrow = nrow(R), ncol = 1)

    # Calculate the difference between R*theta_hat and r
    diff <- R %*% theta_hat - r

    # Calculate the inverse of R*(V_hat/n)*R'
    inv <- solve(R %*% V_hat %*% t(R))

    # Calculate the Wald test statistic
    if (joint_test == "f") {
        wald_statistic <- t(diff) %*% inv %*% diff / dim(R)[1] # Q is the number of rows in R
    } else if (joint_test == "chisq") {
        wald_statistic <- t(diff) %*% inv %*% diff # Not normalized for chi-squared joint_test
    }

    # Degrees of freedom
    if (is.null(df)) {
        df1 <- dim(R)[1] # Q

        if (joint_test == "f") {
            # Check for lme models and warn about df heuristics
            if (inherits(object, "lme") || (!is.null(mfx) && inherits(mfx@model, "lme"))) {
                model_class <- if (inherits(object, "lme")) class(object)[1] else class(mfx@model)[1]
                msg <- "The `hypotheses()` functions uses simple heuristics to select degrees of freedom for this test. See the relevant section in `?hypotheses`. These rules are likely to yield inappropriate results for models of class `%s`. Please supply degrees of freedom values explicitly via the `df` argument."
                warn_sprintf(msg, model_class)
            }

            df2 <- tryCatch(
                insight::get_df(mfx@model),
                error = function(e) NULL
            )
            if (is.null(df2)) {
                tryCatch(insight::get_df(object), error = function(e) NULL)
            }
            if (is.null(df2)) {
                # n: sample size
                n <- tryCatch(stats::nobs(object), error = function(e) NULL)
                if (is.null(n)) {
                    n <- tryCatch(
                        stats::nobs(mfx@model),
                        error = function(e) NULL
                    )
                }
                if (is.null(n)) {
                    stop_sprintf(
                        "Could not extract sample size from model object."
                    )
                }

                df2 <- n - length(theta_hat) # n - P
            }
        }
    } else {
        df1 <- df[1]

        if (joint_test == "f") {
            df2 <- df[2]
        }
    }

    # Calculate the p-value
    if (joint_test == "f") {
        p_value <- 1 - stats::pf(wald_statistic, df1, df2)
    } else if (joint_test == "chisq") {
        p_value <- 1 - stats::pchisq(wald_statistic, df1)
        df2 <- NULL
    }

    # Return the Wald joint_test statistic and p-value
    out <- data.frame(statistic = drop(wald_statistic), p.value = drop(p_value))
    class(out) <- c("hypotheses", "data.frame")
    if (joint_test == "f") {
        attr(out, "statistic_label") <- "F"
    } else if (joint_test == "chisq") {
        attr(out, "statistic_label") <- "ChiSq"
    }

    # degrees of freedom print
    if (joint_test == "f") {
        out$df1 <- df1
        out$df2 <- df2
    } else {
        out$df <- df1
    }

    # Create the print_head string
    print_head <- "\nJoint hypothesis test:\n"
    if (is.character(joint_index)) {
        for (i in joint_index) {
            print_head <- paste0(print_head, i, sprintf(" = %s\n", hypothesis))
        }
    } else if (
        inherits(
            object,
            c("marginaleffects", "comparisons", "slopes", "marginal_means")
        )
    ) {
        tmp <- paste0(
            get_labels(object, idx = joint_index),
            sprintf(" = %s\n", hypothesis)
        )
        print_head <- c(print_head, tmp)
    } else {
        tmp <- paste0(
            get_labels(stats::coef(object), idx = joint_index),
            sprintf(" = %s\n", hypothesis)
        )
        print_head <- c(print_head, tmp)
    }
    attr(out, "print_head") <- print_head

    return(out)
}

File: python/marginaleffects/hypotheses_joint.py
import re
import numpy as np
import scipy.stats as stats
import polars as pl

from .sanity import sanitize_hypothesis_null
from .result import MarginaleffectsResult


def joint_hypotheses(obj, joint_index=None, joint_test="f", hypothesis=0):
    assert joint_test in ["f", "chisq"], "`joint_test` must be `f` or `chisq`"

    # theta_hat: P x 1 vector of estimated parameters
    theta_hat = obj.get_coef()

    var_names = obj.find_predictors()

    if len(theta_hat) == len(var_names) + 1:
        var_names = ["Intercept"] + var_names

    if isinstance(joint_index, bool):
        joint_index = range(len(theta_hat))
    # if joint_index is a string, use grep to find the indices
    elif isinstance(joint_index, str):
        joint_index = [
            i for i in range(len(var_names)) if re.search(joint_index, var_names[i])
        ]
    else:
        if not isinstance(joint_index, list):
            joint_index = [joint_index]
        if all(isinstance(i, str) for i in joint_index):
            joint_index = [
                i for i in range(len(var_names)) if var_names[i] in joint_index
            ]
        assert min(joint_index) >= 0 and max(joint_index) <= len(var_names), (
            "`joint_index` contain invalid indices"
        )

    V_hat = obj.get_vcov()

    # R: Q x P matrix for testing Q hypotheses on P parameters
    # build R matrix based on joint_index
    R = np.zeros((len(joint_index), len(theta_hat)))
    for i in range(len(joint_index)):
        R[i, joint_index[i]] = 1

    if not isinstance(hypothesis, list):
        if hypothesis is None:
            hypothesis = 0
        hypothesis = np.ones(R.shape[0]) * hypothesis
    hypothesis = [sanitize_hypothesis_null(h) for h in hypothesis]

    # Calculate the difference between R*theta_hat and r
    diff = R @ theta_hat - hypothesis

    # Calculate the inverse of R*(V_hat/n)*R'
    inv = np.linalg.inv(R @ V_hat @ R.T)

    # Calculate the Wald test statistic
    if joint_test == "f":
        wald_statistic = (
            diff.T @ inv @ diff / R.shape[0]
        )  # Q is the number of rows in R
    elif joint_test == "chisq":
        wald_statistic = (
            diff.T @ inv @ diff
        )  # Not normalized for chi-squared joint_test

    # Degrees of freedom
    df1 = R.shape[0]  # Q
    df2 = obj.get_df()  # n - P

    # Calculate the p-value
    if joint_test == "f":
        p_value = 1 - stats.f.cdf(wald_statistic, df1, df2)
    elif joint_test == "chisq":
        p_value = 1 - stats.chi2.cdf(wald_statistic, df1)
        df2 = None

    # Return the Wald joint_test statistic and p-value
    mapping = {}
    if joint_test == "f":
        out = pl.DataFrame({"statistic": wald_statistic})
        mapping["statistic"] = "F"
        mapping["p_value"] = "P(>|F|)"
    elif joint_test == "chisq":
        out = pl.DataFrame({"statistic": wald_statistic})
        mapping["statistic"] = "ChiSq"
        mapping["p_value"] = "P(>|ChiSq|)"
    out = out.with_columns(p_value=p_value)

    # degrees of freedom print
    if joint_test == "f":
        out = out.with_columns(df1=df1, df2=df2)
        out = out.cast({"df1": pl.Int64, "df2": pl.Int64})
        mapping["df1"] = "Df 1"
        mapping["df2"] = "Df 2"
    elif joint_test == "chisq":
        out = out.with_columns(df=df1)
        out = out.cast({"df": pl.Int64})
        mapping["df"] = "Df"

    # Create the print_head string
    print_head = "Joint hypothesis test:\n"
    for i, j in enumerate(joint_index):
        print_head += var_names[j] + f" = {hypothesis[i]}\n"
    print_head += "\n"

    out = MarginaleffectsResult(out, mapping=mapping, print_head=print_head)

    return out
